go-controller/pkg/node/egressfirewall_dns/egressfirewall_dns.go

```diff
@@ -0,0 +1,278 @@
+	lock sync.Mutex
+	// holds DNS entries globally
+	dns *util.DNS
+	// name of the node
+	nodeName string
+	// this map holds dnsNames to the dnsEntries
+	dnsEntries map[string]*dnsEntry
+	// allows to get object using the informer chace
+	wf factory.NodeWatchFactory
+	k  kube.Interface
+
+	// Report change when Add operation is done
+	added          chan string
+	removed        chan string
+	stopChan       chan struct{}
+	controllerStop <-chan struct{}
+}
+
+type dnsEntry struct {
+	// the current IP addresses the dnsName resolves to
+	dnsResolves []net.IP
+	namespaces  map[string]struct{} //the namespaces that this DNSName applies to
+}
+
+func NewEgressDNS(nodeName string, watchFactory factory.NodeWatchFactory, k kube.Interface, controllerStop <-chan struct{}) (*EgressDNS, error) {
+	dnsInfo, err := util.NewDNS("/etc/resolv.conf")
+	if err != nil {
+		return nil, err
+	}
+
+	egressDNS := &EgressDNS{
+		dns:        dnsInfo,
+		nodeName:   nodeName,
+		dnsEntries: make(map[string]*dnsEntry),
+		wf:         watchFactory,
+		k:          k,
+
+		added:          make(chan string, 1),
+		removed:        make(chan string, 1),
+		stopChan:       make(chan struct{}),
+		controllerStop: controllerStop,
+	}
+
+	return egressDNS, nil
+}
+
+func (e *EgressDNS) Add(dnsNames []string, namespace string) error {
+	e.lock.Lock()
+	defer e.lock.Unlock()
+	for _, dnsName := range dnsNames {
+		if _, exists := e.dnsEntries[dnsName]; !exists {
+			e.dnsEntries[dnsName] = &dnsEntry{
+				namespaces: map[string]struct{}{
+					namespace: {},
+				},
+			}
+			e.signalAdded(dnsName)
+		} else {
+			// only need to add the namespace to the struct
+			e.dnsEntries[dnsName].namespaces[namespace] = struct{}{}
+		}
+	}
+	return nil
+
+}
+
+//updateEntryForName takes a dnsName and creates/updates the EgressDNS's
+// dnsEntries map for the given dnsName
+func (e *EgressDNS) updateEntryForName(dnsName string) error {
+	e.lock.Lock()
+	defer e.lock.Unlock()
+	ips := e.dns.GetIPs(dnsName)
+	//TODO: if the old and new resolutions are the same should not update the nodeDNSInfo
+	e.dnsEntries[dnsName].dnsResolves = ips
+
+	//update the nodeDNSInfo object so the master can coordinate all the nodes ip addresses
+	// if the node has not reported any DNS addresses before create the nodeDNSObject
+	nodeDNSInfo, err := e.wf.GetNodeDNSInfo(e.nodeName)
+	existed := true
+	if errors.IsNotFound(err) {
+		// the object *MOST LIKELY* has not been created yet, so start making it
+		nodeDNSInfo = &nodednsinfoapi.NodeDNSInfo{
+			ObjectMeta: metav1.ObjectMeta{Name: e.nodeName},
+			Spec:       nodednsinfoapi.NodeDNSInfoSpec{},
+		}
+		existed = false
+	} else if err != nil {
+		return err
+	}
+	var ipStrings []string
+	for _, ip := range e.dnsEntries[dnsName].dnsResolves {
+		ipStrings = append(ipStrings, ip.String())
+	}
+
+	if nodeDNSInfo.Spec.DNSEntries == nil {
+		nodeDNSInfo.Spec.DNSEntries = make(map[string]nodednsinfo.DNSEntry)
+	}
+
+	dnsEntry := nodeDNSInfo.Spec.DNSEntries[dnsName]
```

Comment:
This can be replaced with a single line of code:
```
nodeDNSInfo.Spec.DNSEntries[dnsName] = (whatever the type name is){ IPAddresses: ipStrings}
```